lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -534,12 +591,22 @@ void SingleExitLoopTransformer::ensureSingleExitBlock() {
         if (blocksToBeMoved.count(succ) > 0 || succ == nearestCommonPD) {
           continue;
         }
+        // Clone `succ` if it is reachable by a node outside of this loop.
         if (!DI->properlyDominates(header, succ)) {
-          // Split this edge so that we don't mess up arguments passed in
-          // from other predecessors of succ.
-          if (succ->getNumArguments() > 0) {
-            splitEdge(current->getTerminator(), edgeIdx, DI, LI);
+          // Backedge of a parent loop? This can happen when we have labeled
+          // breaks in loops. We cannot clone the blocks in this case.
+          if (DI->properlyDominates(succ, header)) {
+            // Split this edge so that we don't mess up arguments passed in
+            // from other predecessors of succ.
+            if (succ->getNumArguments() > 0) {
+              splitEdge(current->getTerminator(), edgeIdx, DI, LI);
+            }
+            continue;
           }
+          SILBasicBlock *clonedSucc = cloner.cloneBlock(succ);
+          changeBranchTarget(current->getTerminator(), edgeIdx, clonedSucc,
+                             /*preserveArgs*/ true);
+          worklist.insert(clonedSucc);
           continue;
         }
         worklist.insert(succ);
```

Comment:
the code might be easier to read due to reduced indentation, if we negate "if (!DI->properlyDominates(header, succ))".